File: src/cz/vutbr/fit/interlockSim/sim/SimulationException.java
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.sim;

import jDisco.Process;

/**
 *
 */
public class SimulationException extends Exception {
	private final Object object;
	private final double time;
	/**
	 *
	 */
	public SimulationException() {
		this((Object)null);
	}

	/**
	 * @param object
	 *
	 */
	public SimulationException(Object object) {
		this("", object);
	}

	/**
	 * @param message
	 */
	public SimulationException(String message) {
		this(message, null);
	}

	/**
	 * @param message
	 * @param object
	 */
	public SimulationException(String message, Object object) {
		this(message, null, object);
	}

	/**
	 * @param cause
	 */
	public SimulationException(Throwable cause) {
		this(cause, null);
	}

	/**
	 * @param cause
	 * @param object
	 */
	public SimulationException(Throwable cause, Object object) {
		this("", cause, object);
	}

	/**
	 * @param message
	 * @param cause
	 * @param object
	 */
	public SimulationException(String message, Throwable cause, Object object) {
		super(message, cause);
		this.object = object;
		this.time = Process.time();
	}

	/**
	 * @return object getter
	 */
	public Object getObject() {
		return object;
	}

	/**
	 * @return model time of exception
	 */
	public double getTime() {
		return time;
	}

	@Override
	public String getMessage() {
		return getTime() + ":" + String.valueOf(getObject()) + " " + super.getMessage();
	}
}

File: src/main/kotlin/cz/vutbr/fit/interlockSim/sim/SimulationException.kt
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.sim

import jDisco.Process

/**
 *
 */
open class SimulationException : Exception {
	private val obj: Any?
	private val time: Double

	/**
	 *
	 */
	constructor() : this(null as Any?)

	/**
	 * @param object
	 *
	 */
	constructor(obj: Any?) : this("", obj)

	/**
	 * @param message
	 */
	constructor(message: String) : this(message, null as Any?)

	/**
	 * @param message
	 * @param object
	 */
	constructor(message: String, obj: Any?) : this(message, null as Throwable?, obj)

	/**
	 * @param cause
	 */
	constructor(cause: Throwable?) : this(cause, null as Any?)

	/**
	 * @param cause
	 * @param object
	 */
	constructor(cause: Throwable?, obj: Any?) : this("", cause, obj)

	/**
	 * @param message
	 * @param cause
	 * @param object
	 */
	constructor(message: String, cause: Throwable?, obj: Any?) : super(message, cause) {
		this.obj = obj
		this.time = Process.time()
	}

	/**
	 * @return object getter
	 */
	open fun getObject(): Any? = obj

	/**
	 * @return model time of exception
	 */
	fun getTime(): Double = time

	override fun toString(): String {
		val msg = message?.takeIf { it.isNotEmpty() } ?: ""
		return "${this::class.simpleName}: $msg at time $time"
	}
}
